lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp

```diff
@@ -0,0 +1,330 @@
+//===- MergeConnections.cpp - Merge expanded connections --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass merges expanded connections into one connection.
+// LowerTypes fully expands aggregate connections even when semantically
+// not necessary to expand because it is required for ExpandWhen.
+//
+// More specifically this pass folds the following patterns:
+//   %dest(0) <= subfield %a(0)
+//   %dest(1) <= subfield %a(1)
+//   ...
+//   %dest(n) <= subfield %a(n)
+// into
+//   %dest <= %a
+// This pass also merge connections if all of their source values are known to
+// be constant.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include <queue>
+
+#define DEBUG_TYPE "firrtl-merge-connections"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+// Return ture if it might be fine to merge the connect.
+static bool okToPeel(ConnectOp connect) {
+  // Ignore connections between different types.
+  if (connect.src().getType() != connect.dest().getType())
```

Comment:
This isn't strictly necessary, right?  connects don't need pairwise identical types.